File: BOJ/Shortest_path/BOJ_11657.py
import sys
input = sys.stdin.readline

INF = int(1e9)
Flag = False

N, M = map(int, input().split())
graph = [[] for _ in range(N+1)]

for i in range(M):
    A, B, C = map(int, input().split())
    graph[A].append((B, C))

distance = [INF] * (N+1)

distance[1] = 0

for i in range(1, N+1):
    for j in range(1, N+1):
        for k in graph[j]:
            new_city = k[0]
            dist = k[1]
            if distance[j] != INF and distance[new_city] > distance[j] + dist :
                distance[new_city] = distance[j] + dist

    if i == N:
        for j in range(1, N+1):
            for k in graph[j]:
                new_city = k[0]
                dist = k[1]
                if distance[j] != INF and distance[new_city] > distance[j] + dist :
                    Flag = True

if Flag:
    print(-1)
else :
    for i in range(2, N+1):
        if distance[i] == INF:
            print(-1)
        else :
            print(distance[i])
    

File: BOJ/Shortest_path/BOJ_11657.cpp
#include <iostream>
#include <vector>

#define NMAX 501

using namespace std;

const int INF = 1e9;
struct Node{
    int from, to, w;
};

long long dist[NMAX];

int main(){
    int N, M;
    cin >> N >> M ;

    for (int i = 0; i < N+1; i++){
        dist[i] = INF;
    }
    vector<Node> graph;

    for (int i = 0; i < M; i++){
        int A, B, C;
        cin >> A >> B >> C;
        graph.push_back({A, B, C});
    }

    dist[1] = 0;

    for (int i = 1; i < N+1; i++){
        for (int j = 0; j < M; j++){
            int from = graph[j].from, to = graph[j].to, cost = graph[j].w;

            if ((dist[from] != INF) && (dist[to] > dist[from] + cost)){
                dist[to] = dist[from] + cost;
                if (i == N) {
                    cout << -1 << endl;
                    return 0;
                }
            }
        }
    }

    for (int i = 2; i < N+1; i++){
        if (dist[i] == INF) cout << -1 << endl;
        else cout << dist[i] << endl;
    }

    return 0;
}